### ISSUE TITLE
Optimize speed inserting into DB

### ISSUE THREAD

**Issue by michaelsilver** [infolab-csail/wikithingsdb#6]
As of commit f92fb9eca384952ccdb0d96613615d273f32c823, it takes 4 minutes to insert 678 articles — 164 days to load 40,000,000 articles if I did the calculation correctly.

I was thinking of using a combo of http://docs.sqlalchemy.org/en/rel_1_0/orm/tutorial.html#working-with-related-objects (using an array to insert relationships), what @alvaromorales  mentioned by using http://docs.sqlalchemy.org/en/rel_1_0/orm/persistence_techniques.html#bulk-operations, and also using the commit every 1000 or so additions.

**Comment by alvaromorales** [comment]
Perhaps the bottleneck is whoami. Can you time how long it takes to insert 678 articles if you don't get the types from whoami?

**Comment by alvaromorales** [comment]
Note that there are only around ~5M Wikipedia **articles**. Articles have page_namespace=0. More about namespaces: https://www.mediawiki.org/wiki/Manual:Namespace#Built-in_namespaces

For elasticstart, I additionally skip some articles:

```python
def ignore_page(data):
    # Using string contains because it's faster than regex search
    disambig_pattern = '(disambiguation)'
    listof_pattern = 'List of'
    referto_pattern = 'may refer to'

    title = data['title']

    is_disambig = disambig_pattern in title
    is_list = listof_pattern in title
    is_refer = referto_pattern in data['abstract']

    return is_disambig or is_list or is_refer
```

**Comment by michaelsilver** [comment]
> Perhaps the bottleneck is whoami. Can you time how long it takes to insert 678 articles if you don't get the types from whoami?

Yup, if I skip making a request to whoami and just give a fake list to insert (e.g. `return ["foo", "bar", "baz"]` as types), it takes 11 seconds to insert 678 articles --- which translates to under an hour to insert 11M articles. @alvaromorales, is there any way to set up Whoami as an importable library that can do everything natively without the need of a request server?

**Comment by alvaromorales** [comment]
No, but whoami is easily parallelizable. I think you should focus on getting the other parts working. I can then launch multiple whoami servers, and make a PR that uses multiprocessing for whoami and solves #5, very similar to what I already did for elasticstart.

**Comment by michaelsilver** [comment]
Could you please open a PR for this? -- I think we are at the point where it is appropriate, and it will certainly make the "sanity checks" you mentioned in PR #9 easier to implement. As of now, I am limited by only having access to full info about 678 articles, most of which are on obscure topics.

When you do it, if you could include comments or some sort of explanation since I am interesting in learning about multiprocessing, that would be awesome!

**Comment by michaelsilver** [comment]
Closed by #10
